filesystem/flysystem/FlysystemAssetStore.php

```diff
@@ -0,0 +1,271 @@
+<?php
+
+namespace SilverStripe\Filesystem\Flysystem;
+
+use Config;
+use Injector;
+use InvalidArgumentException;
+use League\Flysystem\Exception;
+use League\Flysystem\Filesystem;
+use SilverStripe\Filesystem\Storage\AssetNameGenerator;
+use SilverStripe\Filesystem\Storage\AssetStore;
+
+/**
+ * Asset store based on flysystem Filesystem as a backend
+ */
+class FlysystemAssetStore implements AssetStore {
+
+	/**
+	 * @var Filesystem
+	 */
+	private $filesystem = null;
+
+	/**
+	 * Enable to use legacy filename behaviour (omits hash)
+	 *
+	 * @config
+	 * @var bool
+	 */
+	private static $legacy_filenames = false;
+
+	/**
+	 * Assign new flysystem backend
+	 *
+	 * @param Filesystem $filesystem
+	 * @return $this
+	 */
+	public function setFilesystem(Filesystem $filesystem) {
+		$this->filesystem = $filesystem;
+		return $this;
+	}
+
+	/**
+	 * Get the currently assigned flysystem backend
+	 *
+	 * @return Filesystem
+	 */
+	public function getFilesystem() {
+		return $this->filesystem;
+	}
+
+	public function getAsStream($hash, $filename, $variant = null) {
+		$fileID = $this->getFileID($hash, $filename, $variant);
+		return $this->getFilesystem()->readStream($fileID);
+	}
+
+	public function getAsString($hash, $filename, $variant = null) {
+		$fileID = $this->getFileID($hash, $filename, $variant);
+		return $this->getFilesystem()->read($fileID);
+	}
+
+	public function getAsURL($hash, $filename, $variant = null) {
+		$fileID = $this->getFileID($hash, $filename, $variant);
+		return $this->getFilesystem()->getPublicUrl($fileID);
+	}
+
+	public function setFromLocalFile($path, $filename = null, $conflictResolution = null) {
+		// Get filename to save to
+		if(empty($filename)) {
+			$filename = basename($path);
+		}
+		
+		// Get local content
+		$content = file_get_contents($path);
+		if($content === false) {
+			throw new InvalidArgumentException("$path does not exist");
+		}
+
+		// Write as string
+		return $this->setFromString($content, $filename, $conflictResolution);
+	}
+
+	public function setFromStream($stream, $filename, $conflictResolution = null) {
+		$content = stream_get_contents($stream);
```

Comment:
FYI: I just did a quick test: running `sha1_file()` on a 2.1GB file executed without incident in 5.8s and `memory_get_peak_usage()` return 262kb.